File: examples/python/p256-encode-negate/P256EncodeNegate.runar.py
"""P256EncodeNegate -- Python port. Executed coverage for p256_negate and
p256_encode_compressed, neither of which appeared in any fixture.
See the `.runar.ts` port for the full rationale.
"""

from runar import SmartContract, ByteString, public, assert_, p256_negate, p256_encode_compressed


class P256EncodeNegate(SmartContract):
    expected_compressed: ByteString

    def __init__(self, expected_compressed: ByteString):
        super().__init__(expected_compressed)
        self.expected_compressed = expected_compressed

    @public
    def check_negate(self, p: P256Point, expected: P256Point):
        """(x, y) -> (x, p - y). Guards canonicity AND the 64-byte width."""
        n = p256_negate(p)
        assert_(n == expected)

    @public
    def check_encode(self, p: P256Point, expected: ByteString):
        """Point -> 33-byte 02/03||x, parity read at a fixed offset."""
        e = p256_encode_compressed(p)
        assert_(e == expected)

    @public
    def check_negate_then_encode(self, p: P256Point):
        """Composed: compressing the negation must flip the prefix only."""
        n = p256_negate(p)
        e = p256_encode_compressed(n)
        assert_(e == self.expected_compressed)

File: examples/zig/p256-encode-negate/P256EncodeNegate.runar.zig
const runar = @import("runar");

// P256EncodeNegate -- Zig port. Executed coverage for p256Negate and
// p256EncodeCompressed, neither of which appeared in any fixture.
// See the `.runar.ts` port for the full rationale.
pub const P256EncodeNegate = struct {
    pub const Contract = runar.SmartContract;

    expectedCompressed: runar.ByteString,

    pub fn init(expectedCompressed: runar.ByteString) P256EncodeNegate {
        return .{ .expectedCompressed = expectedCompressed };
    }

    /// (x, y) -> (x, p - y). Guards coordinate canonicity AND the 64-byte width.
    pub fn checkNegate(self: *const P256EncodeNegate, p: runar.P256Point, expected: runar.P256Point) void {
        _ = self;
        const n = runar.p256Negate(p);
        runar.assert(n == expected);
    }

    /// Point -> 33-byte 02/03||x. Guards the width; parity read at a fixed offset.
    pub fn checkEncode(self: *const P256EncodeNegate, p: runar.P256Point, expected: runar.ByteString) void {
        _ = self;
        const e = runar.p256EncodeCompressed(p);
        runar.assert(runar.bytesEq(e, expected));
    }

    /// Composed: compressing the negation must flip the prefix and nothing else.
    pub fn checkNegateThenEncode(self: *const P256EncodeNegate, p: runar.P256Point) void {
        const n = runar.p256Negate(p);
        const e = runar.p256EncodeCompressed(n);
        runar.assert(runar.bytesEq(e, self.expectedCompressed));
    }
};
